### PR TITLE
Added advanced code generation

### PR DESCRIPTION
**Pull request by adranwit**
- dependent type included in the generated code
-  embedded type support
-  referring type support (i.e type Payload []byte)
-  option to generate code reusing object with sync.Pool
-  custom annotation tag support, i.e   `"json:"key1"`  vs `gojay:"key1"`
-  time formatting support

### CODE REVIEW

**Comment by adranwit** [comment]
Waiting for @francoispqt  to either reject it or accept,  the code generation uses based classes provided by this project, so I believe this is a good place for it.

**Comment by francoispqt** [comment]
Sorry for the delay to review the PR. It’s been hanging out here way too long. I’m reviewing/testing it right now.

File: codegen/options.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	Types       []string
+	PoolObjects bool
+	TagName     string
+}
+
+func (o *Options) Validate() error {
+	if o.Source == "" {
+		return errors.New("Source was empty")
+	}
+	if o.Dest == "" {
```
Comment (francoispqt) [inline codegen/options.go]: I'm removing this, it prevents from printing to stdout if no "o" flag is provided.

**Comment by francoispqt** [comment]
So there are nice things. Overall the design is better than the original codegen tool. I like also the possibility to pool objects. 

I'm removing the binary in the PR, fixed the bug for stdout output, update dep files and will merge to master. 

I will let both gen tools live here for now but will replace the old one with this one after:
- add support for a bunch of types that were added like `sql.NullXxx` 
- add support for `omitempty` and `nullempty`

Thank you

File: codegen/options.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		return errors.New("Dest was empty")
+	}
```
Comment (francoispqt) [inline codegen/options.go]: I'm removing this, it prevents from printing to stdout if no "o" flag is provided.

**Pull request merged**